Bump rocksdb from v8.3.2 to v8.3.3 and change the default value of rocksdb.level_compaction_dynamic_level_bytes from no to yes

aleksraiden opened the pull request:
This is a bugfix release in 8.3 line (8.6.х are development now).

**Bugfix**

- Fix a bug where if there is an error reading from offset 0 of a file from L1+ and that the file is not the first file in the sorted run, data can be lost in compaction and read/scan can return incorrect results.
- Fix a bug where iterator may return incorrect result for DeleteRange() users if there was an error reading from a file.
- Fixed a race condition in GenericRateLimiter that could cause it to stop granting requests

**Important**! Config options level_compaction_dynamic_level_bytes defaults changed to true in kvrocks.config file

Original release notice: https://github.com/facebook/rocksdb/releases/tag/v8.3.3

aleksraiden committed:
Update rocksdb.cmake

mapleFU commented:
The bug fixes LGTM

Also I notice lots of behavior changes and some new API/changes about compaction in 8.4 (Notice that some are DeleteRange related, some are new statistics)

`level_compaction_dynamic_level_bytes` is default to `true`, and it's false in kvrocks config. Should we try to switch that? Or keep the current behavior?

aleksraiden commented:
Thanks, @mapleFU!  I think, we can change our defaults to rocksdb defaults. I can add in this patch, if other reviewers are OK.

mapleFU commented:
Yes. All these bugfixes LGTM. But there are some changes over compaction, I think having a benchmark or others would be better

aleksraiden committed:
Merge branch 'unstable' into rocksdb-8.5.3

git-hulk committed:
Merge branch 'unstable' into rocksdb-8.5.3

git-hulk commented:
@aleksraiden @mapleFU How about upgrading to v8.3.3 first? since those critical fixes are also fixed in this version. And then we can have more tests before upgrading to the next minor version.

https://github.com/facebook/rocksdb/releases/tag/v8.3.3

## Bug Fixes
- Fix a bug where if there is an error reading from offset 0 of a file from L1+ and that the file is not the first file in the sorted run, data can be lost in compaction and read/scan can return incorrect results.
- Fix a bug where iterator may return incorrect result for DeleteRange() users if there was an error reading from a file.
- Fixed a race condition in GenericRateLimiter that could cause it to stop granting requests

PragmaTwice commented:
Maybe we can upgrade to 8.3.3 before releasing our 2.6.0, and then upgrade to 8.5 in the unstable branch.

OFF TOPIC: I think we now need a release manager for 2.6.0.

git-hulk commented:
> OFF TOPIC: I think we now need a release manager for 2.6.0.

@apache/kvrocks-committers Not sure if any committer has time to drive this process recently? Can refer [how to create a Kvrocks release](https://kvrocks.apache.org/community/create-a-release)

mapleFU commented:
Yeah, I'm not an expert on RocksDB, so I don't know how changes in `8.5.3` will affect kvrocks, though it doesn't meens we're locked in 8.3, but I think we should take upgrade carefully... 

But `8.3.3` is always ok to me...

git-hulk commented:
> Yeah, I'm not an expert on RocksDB, so I don't know how changes in 8.5.3 will affect kvrocks, though it doesn't meens we're locked in 8.3, but I think we should take upgrade carefully...

Agree we shouldn't be too aggressive when it's related to storage(rocksdb), we need more tests before upgrading to the minor version since it may introduce some unexpected behaviors.

caipengbo commented:
Yes, I also think we upgrade the underlying engine is too radical. I agreed to upgrade to 8.3.3, one of the bugs about `GenericRateLimiter` we encountered in production and I reported it(https://github.com/facebook/rocksdb/issues/11742).

Like @mapleFU suggested, we should set `level_compaction_dynamic_level_bytes` to true in kvrocks, this has a lot of benefits, and [rocksdb8.2.0](https://github.com/facebook/rocksdb/blob/main/HISTORY.md#820-04242023) can be easily to migrate `level_compaction_dynamic_level_bytes` from false to true( https://github.com/facebook/rocksdb/pull/11321)

aleksraiden committed:
Update rocksdb.cmake

Revert to rocksdb 8.3.3

aleksraiden committed:
Update kvrocks.conf

Set level_compaction_dynamic_level_bytes defaults to true (see https://github.com/facebook/rocksdb/pull/11321)

aleksraiden commented:
Okey, many thanks about discussion. I changed PR to support only 8.3.3 and send new PR in next release. Also, config options level_compaction_dynamic_level_bytes changed to true.

enjoy-binbin committed:
Merge branch 'unstable' into rocksdb-8.5.3
